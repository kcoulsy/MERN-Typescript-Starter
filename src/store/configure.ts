import { configureStore } from '@reduxjs/toolkit';

// Import Reducers
import authReducer from '../reducers/auth';

const store = configureStore({
    reducer: {
        // reducers here
        auth: authReducer,
    },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
