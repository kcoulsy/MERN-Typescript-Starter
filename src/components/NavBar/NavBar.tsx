import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { startLogout } from '../../actions/auth';
import isLoggedIn from '../../helpers/isLoggedIn';
import { RootState } from '../../store/configure';

interface NavBarProps {
    isLoggedIn: boolean;
    startLogout: Function;
}

export const NavBar: React.FunctionComponent<NavBarProps> = ({ isLoggedIn, startLogout }) => {
    const handleLogout = (ev: React.MouseEvent<HTMLAnchorElement>) => {
        ev.preventDefault();
        startLogout();
    };

    return (
        <div className="ui stackable menu">
            <Link to="/" className="header item">
                Home
            </Link>
            <div className="right menu">
                {!isLoggedIn ? (
                    <Fragment>
                        <Link to="/register" className="item">
                            Register
                        </Link>
                        <Link to="/login" className="item">
                            Login
                        </Link>
                    </Fragment>
                ) : (
                    <Fragment>
                        <Link to="/secret" className="item">
                            Secret
                        </Link>
                        <a href="/" className="item" id="logout" onClick={handleLogout}>
                            Logout
                        </a>
                    </Fragment>
                )}
            </div>
        </div>
    );
}

export const mapStateToProps = (state: RootState) => {
    return {
        isLoggedIn: isLoggedIn(state.auth.token),
    };
};

const mapDispatchToProps = { startLogout };

export default connect(mapStateToProps, mapDispatchToProps)(NavBar);
