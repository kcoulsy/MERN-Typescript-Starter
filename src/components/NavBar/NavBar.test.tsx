import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { NavBar, mapStateToProps } from './NavBar';
import { RootState } from '../../store/configure';

describe('NavBar component tests', () => {
    it('should render correctly', () => {
        const startLogout = jest.fn();
        const wrapper = shallow(<NavBar isLoggedIn={false} startLogout={startLogout} />);

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('shoulder render log in and logout if not logged in', () => {
        const startLogout = jest.fn();
        const wrapper = shallow(<NavBar startLogout={startLogout} isLoggedIn={false} />);

        expect(wrapper.find('Link[to="/login"]').length).toBe(1);
        expect(wrapper.find('Link[to="/register"]').length).toBe(1);
        expect(wrapper.find('Link[to="/secret"]').length).toBe(0);
        expect(wrapper.find('a[id="logout"]').length).toBe(0);

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('shoulder render logout if logged in', () => {
        const startLogout = jest.fn();
        const wrapper = shallow(<NavBar startLogout={startLogout} isLoggedIn={true} />);

        expect(wrapper.find('Link[to="/login"]').length).toBe(0);
        expect(wrapper.find('Link[to="/register"]').length).toBe(0);
        expect(wrapper.find('Link[to="/secret"]').length).toBe(1);
        expect(wrapper.find('a[id="logout"]').length).toBe(1);

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('clicking logout should log out', () => {
        const startLogout = jest.fn();
        const wrapper = shallow(<NavBar startLogout={startLogout} isLoggedIn={true} />);

        expect(wrapper.find('Link[to="/login"]').length).toBe(0);
        expect(wrapper.find('Link[to="/register"]').length).toBe(0);
        expect(wrapper.find('Link[to="/secret"]').length).toBe(1);
        expect(wrapper.find('a[id="logout"]').length).toBe(1);

        wrapper.find('a[id="logout"]').simulate('click', { preventDefault: jest.fn() });

        expect(startLogout.mock.calls.length).toBe(1);

        wrapper.setProps({ isLoggedIn: false });

        expect(wrapper.find('Link[to="/login"]').length).toBe(1);
        expect(wrapper.find('Link[to="/register"]').length).toBe(1);
        expect(wrapper.find('Link[to="/secret"]').length).toBe(0);
        expect(wrapper.find('a[id="logout"]').length).toBe(0);
    });

    it('mapStateToProps should return isLoggedIn prop', () => {
        const mockState: RootState = {
            auth: {
                attemptingLogin: false,
                token: 'testtoken',
                loginAttemptFailed: false,
                appInitialising: false,
                appInitialised: false,
                attemptingRegister: false,
            },
        };

        expect(mapStateToProps(mockState).isLoggedIn).toBe(true);
    });
});
