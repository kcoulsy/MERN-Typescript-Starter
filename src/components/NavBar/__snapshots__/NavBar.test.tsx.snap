// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar component tests should render correctly 1`] = `
<div
  className="ui stackable menu"
>
  <Link
    className="header item"
    to="/"
  >
    Home
  </Link>
  <div
    className="right menu"
  >
    <Link
      className="item"
      to="/register"
    >
      Register
    </Link>
    <Link
      className="item"
      to="/login"
    >
      Login
    </Link>
  </div>
</div>
`;

exports[`NavBar component tests shoulder render log in and logout if not logged in 1`] = `
<div
  className="ui stackable menu"
>
  <Link
    className="header item"
    to="/"
  >
    Home
  </Link>
  <div
    className="right menu"
  >
    <Link
      className="item"
      to="/register"
    >
      Register
    </Link>
    <Link
      className="item"
      to="/login"
    >
      Login
    </Link>
  </div>
</div>
`;

exports[`NavBar component tests shoulder render logout if logged in 1`] = `
<div
  className="ui stackable menu"
>
  <Link
    className="header item"
    to="/"
  >
    Home
  </Link>
  <div
    className="right menu"
  >
    <Link
      className="item"
      to="/secret"
    >
      Secret
    </Link>
    <a
      className="item"
      href="/"
      id="logout"
      onClick={[Function]}
    >
      Logout
    </a>
  </div>
</div>
`;
